### PR TITLE
[SYCL] Add new SYCL attribute readonly

### PR DESCRIPTION
**Pull request by elizabethandrews**
If a SYCL accessor is templated on access::mode::read, then represent
this in LLVM IR as the readonly attribute applied on the kernel argument
corresponding to the accessor.

Co-author : Jeffrey Mott <<jeffrey.t.mott@intel.com>>

Signed-off-by: Elizabeth Andrews <<elizabeth.andrews@intel.com>>

### CODE REVIEW

File: clang/lib/Sema/SemaSYCL.cpp
Comment (erichkeane) [inline clang/lib/Sema/SemaSYCL.cpp:1113]: Cant the above just be castAs which should do an assert?

File: clang/lib/Sema/SemaSYCL.cpp
Comment (erichkeane) [inline clang/lib/Sema/SemaSYCL.cpp:1109]: Why the 'out' parameter rather than just returning E->GetInitVal?

File: clang/lib/Sema/SemaSYCL.cpp
Comment (erichkeane) [inline clang/lib/Sema/SemaSYCL.cpp:2144]: oh, definitely just have getAccessModeReadVal return the APSInt here.

File: clang/lib/Sema/SemaSYCL.cpp
Comment (erichkeane) [inline clang/lib/Sema/SemaSYCL.cpp:2145]: in fact, unless there is a call to the `getAccessModeReadVal` that I'm missing, it probably should just return 'bool' for this condition.

File: clang/lib/Sema/SemaSYCL.cpp
Comment (elizabethandrews) [reply clang/lib/Sema/SemaSYCL.cpp]: Returned bool instead.

File: clang/lib/Sema/SemaSYCL.cpp
Comment (elizabethandrews) [reply clang/lib/Sema/SemaSYCL.cpp]: Done

File: clang/lib/Sema/SemaSYCL.cpp
Comment (erichkeane) [inline clang/lib/Sema/SemaSYCL.cpp:1113]: Enum namescan't be duplicated, so you can make this all:
```suggestion
    if (E->getName() == "read") return E->getInitVal() == AccessModeArg.getAsIntegral();
```

**Review by erichkeane — COMMENTED** [review]
1 more comment, otherwise LGTM.

**Review by erichkeane — DISMISSED** [review]
I'm happy, but Aaron and Prem should review.

File: clang/lib/Sema/SemaSYCL.cpp
Comment (AaronBallman) [inline clang/lib/Sema/SemaSYCL.cpp:1114]: `llvm::find_if()`?

**Review by AaronBallman — COMMENTED** [review]
Basically LGTM, just some tiny nits.

**Review by premanandrao — COMMENTED** [review]
LGTM, modulo changes requested by Aaron.

File: clang/lib/Sema/SemaSYCL.cpp
Comment (erichkeane) [inline clang/lib/Sema/SemaSYCL.cpp:1118]: Er... you don't need a ternary to change 1 boolean to another.  
```suggestion
  return IsReadOnly != ED->enumerator_end();
```

File: clang/lib/Sema/SemaSYCL.cpp
Comment (erichkeane) [inline clang/lib/Sema/SemaSYCL.cpp:1113]: FWIW, and I don't really care enough to do more than say it, this ends up being a bit less efficient than finding `read`, then returning whether it matches, since this can't early-exit.

File: clang/lib/Sema/SemaSYCL.cpp
Comment (elizabethandrews) [reply clang/lib/Sema/SemaSYCL.cpp:1118]: Hahaha! I have no idea why I did this. Thanks for catching it

File: clang/lib/Sema/SemaSYCL.cpp
Comment (elizabethandrews) [reply clang/lib/Sema/SemaSYCL.cpp:1113]: Can you please elaborate? Do you mean this won't be efficient if the value of `InitVal `of read and `AccessmodeArg` doesn't match?

Comment (erichkeane) [reply clang/lib/Sema/SemaSYCL.cpp:1113]: Right.  The earlier version searched until it found 'read', then stopped going through the loop.  The new version goes through the loop until it finds 'read' AND `InitVal`.  If the `read` element doesn't match `InitVal` this searches the whole loop.

Comment (elizabethandrews) [reply clang/lib/Sema/SemaSYCL.cpp:1113]: Yes. I can change that if you'd like. The chance of that happening is low though....Unless the library implementer decides to change the value for some reason

Comment (AaronBallman) [reply clang/lib/Sema/SemaSYCL.cpp]: I don't have opinions either way. :-)

File: clang/lib/Sema/SemaSYCL.cpp
Comment (AaronBallman) [inline clang/lib/Sema/SemaSYCL.cpp:1117]:
```suggestion
  return ReadOnly != ED->enumerator_end() &&
```
Nice, this approach also works well!

File: clang/lib/Sema/SemaSYCL.cpp
Comment (AaronBallman) [inline clang/lib/Sema/SemaSYCL.cpp:1111]:
```suggestion
  const EnumDecl *ED = AccessModeArgEnumType->getDecl();
```

File: clang/lib/Sema/SemaSYCL.cpp
Comment (AaronBallman) [inline clang/lib/Sema/SemaSYCL.cpp:1113]:
```suggestion
  auto ReadOnly = llvm::find_if(ED->enumerators(), [&](const EnumConstantDecl *E) {
```

File: clang/lib/Sema/SemaSYCL.cpp
Comment (AaronBallman) [inline clang/lib/Sema/SemaSYCL.cpp:2128]:
```suggestion
    const TemplateArgument &AccessModeArg =
```

File: clang/lib/Sema/SemaSYCL.cpp
Comment (AaronBallman) [inline clang/lib/Sema/SemaSYCL.cpp:2158]:
```suggestion
    const TemplateArgument &AccessModeArg =
```

**Review by premanandrao — APPROVED** [review]
LGTM

**Review by AaronBallman — APPROVED** [review]
LGTM!

**Pull request merged**
